### ISSUE TITLE
`TokenizeContext.sliceSerialize` for `Token.type` of `setextHeading` includes non-heading content from outside the range of [`startLine`, `endLine`]

### ISSUE THREAD

**Issue by DavidAnson** [micromark/micromark#152]
### Initial checklist

- [X] I read the [support docs](https://github.com/micromark/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/micromark/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/micromark/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amicromark&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

4.0.0

### Steps to reproduce

```
user@HOST micromark-setext % npm ls micromark
micromark-setext@ /Users/user/Documents/micromark-setext
└── micromark@4.0.0

user@HOST micromark-setext % cat issue.mjs 
import { parse } from "micromark";
import { postprocess } from "micromark";
import { preprocess } from "micromark";

const markdown = `
Text

Setext
======

Text
`;

const encoding = undefined;
const end = true;
const options = undefined;
const chunks = preprocess()(markdown, encoding, end);
const parseContext = parse(options).document().write(chunks);
const events = postprocess(parseContext);
for (const event of events) {
  const [ kind, token, context ] = event;
  if (kind === "enter") {
    const { type, start, end } = token;
    const { "line": startLine } = start;
    const { "line": endLine } = end;
    console.dir(`${type} (${startLine}-${endLine}): ${context.sliceSerialize(token)}`);
  }
}
user@HOST micromark-setext % node issue.mjs  
'lineEndingBlank (1-2): \n'
'content (2-2): Text'
'paragraph (2-2): Text'
'data (2-2): Text'
'lineEnding (2-3): \n'
'lineEndingBlank (3-4): \n'
'setextHeading (4-5): Text\n\nSetext\n======'
'setextHeadingText (4-4): Setext'
'data (4-4): Setext'
'lineEnding (4-5): \n'
'setextHeadingLine (5-5): ======'
'setextHeadingLineSequence (5-5): ======'
'lineEnding (5-6): \n'
'lineEndingBlank (6-7): \n'
'content (7-7): Text'
'paragraph (7-7): Text'
'data (7-7): Text'
'lineEnding (7-8): \n'
user@HOST micromark-setext %
```

### Expected behavior

Note specifically this part of the output: `'setextHeading (4-5): Text\n\nSetext\n======'`

While the start and end lines are correct, the output of `sliceSerialize` includes "Text\n\n" from lines 2 and 3 which is not part of the heading (confirmed by the associated `setextHeadingText` token which contains only "Setext").

### Actual behavior

See above.

### Runtime

Node v16

### Package manager

npm v7

### OS

macOS

**Comment by wooorm** [comment]
Here’s the code  https://github.com/micromark/micromark/blob/4346e010d537b05f622e636d001bd25c1140f6a1/packages/micromark-core-commonmark/dev/lib/setext-underline.js#L62C1-L62C1.

Might be that `markdown-rs` does have this better. I ported most of its changes over to micromark, but not some of the more complex stuff handling events, which is exactly what has to happen for setext headings

**Comment by wooorm** [comment]
Hi again! I recommend using `offset`s instead of these contexts. The regular positional info (line/column/offset) is heavily tested and used across the ecosystem. You can slice the original document with those offset.

The irregular stuff (`_index`, `_bufferIndex`) exists for virtual spaces that arise because things in markdown can start and end “inside” tabs. Complex things that you probably don’t care about. For performance reasons these internal APIs aren’t given the thought that you expect!

**Comment by DavidAnson** [comment]
@wooorm, thank you for coming back to this! To confirm I understand your suggestion, you are saying that instead of calling `sliceSerialize` to get original Markdown for a token, that I should index into the input string instead? I'm happy to do that, but as I recall (on my phone right now so can't look it up easily), the input I provide is a single string and this trying to rely on line and column number would require my code to exactly match the line-splitting used by the parser (and maybe need to deal with encoding plus escaping issues, too?). Alternatively, I think offset is directly usable, but without a corresponding length property, I run into the same problem of not knowing where to stop. Am I misremembering the situation or is there a reliable way to do what you suggest? Thank you again for the help!

**Comment by DavidAnson** [comment]
Okay, I had a chance to look at the code during a meeting and I see that `offset` is a property of `Point` so the span from `start` to `end` of `Token` should meet my needs. I will try that soon!

**Comment by wooorm** [comment]
Yep, offsets are indices into strings! A lot of the ecosystem here uses them. If, from your earlier comment, you want to translate between things, see: https://github.com/vfile/vfile-location

**Comment by wooorm** [comment]
Released in https://github.com/micromark/micromark/releases/tag/micromark-core-commonmark%402.0.3!
